src/events/utilities/grade-calculator/grade-calculator.ts

```diff
@@ -14,28 +16,31 @@ export class GradeCalculator {
   }
 
   public static computeGradeStream(distanceStream: number[], altitudeStream: number[]): number[] {
+    // First filter the altitude to remove any noise
+    const kf = new KalmanFilter();
+    altitudeStream = altitudeStream.map(v => kf.filter(v));
 
-    let gradeStream = [];
+    const gradeStream = [];
 
     for (let i = 0; i < distanceStream.length; i++) {
 
-      const previousDistance = (distanceStream[i - 1]) ? distanceStream[i - 1] : 0;
-      const currentDistance = distanceStream[i];
-      const previousAltitude = (altitudeStream[i - 1]) ? altitudeStream[i - 1] : altitudeStream[i];
-      const currentAltitude = altitudeStream[i];
+      const previousDistance = (distanceStream[i - 1]) || 0;
+      // Find the next i for distance = prev distance + 5m
+      let nextIndex = distanceStream.slice(i).findIndex(d => d >= (previousDistance + 5));
+      // If not found -1 then make it 0 (current)
+      nextIndex = nextIndex === -1 ? 0 : nextIndex
+      const currentDistance = distanceStream[nextIndex + i];
+      const previousAltitude = isNumber(altitudeStream[i - 1]) ? altitudeStream[i - 1] : altitudeStream[i];
+      const currentAltitude = altitudeStream[nextIndex + i];
 
       const currentGrade = GradeCalculator.computeGrade(previousDistance, currentDistance, previousAltitude, currentAltitude);
       gradeStream.push(currentGrade);
     }
 
-    const lowPassFilter = new LowPassFilter(0.55);
-    gradeStream = lowPassFilter.smoothArray(gradeStream);
-    gradeStream = lowPassFilter.smoothArray(gradeStream);
-    gradeStream.push((gradeStream[gradeStream.length - 1] + gradeStream[gradeStream.length - 2]) / 2); // "Predict last sample before shift"
-    gradeStream.push((gradeStream[gradeStream.length - 1] + gradeStream[gradeStream.length - 2]) / 2); // "Predict last sample before shift"
-    gradeStream.shift();
-    gradeStream.shift();
-    return gradeStream;
+    kf = new KalmanFilter();
```

Comment:
Whoops. Seems to a commit error :)

Reply:
No worries fixed. Perhaps the filter is not needed. We will wipe those out when we agree the tests / fixture prove the thing working :-D